crypto/fipsmodule/rsa/rsa.c

```diff
@@ -448,12 +448,144 @@ int RSA_set0_crt_params(RSA *rsa, BIGNUM *dmp1, BIGNUM *dmq1, BIGNUM *iqmp) {
+}
+
+int RSA_set_method(RSA *rsa, const RSA_METHOD *meth) {
+  if(rsa == NULL || meth == NULL) {
+    OPENSSL_PUT_ERROR(RSA, ERR_R_PASSED_NULL_PARAMETER);
+    return 0;
+  }
+
+  rsa->meth = meth;
+  return 1;
+}
+
+const RSA_METHOD *RSA_get_method(const RSA *rsa) {
+  if(rsa == NULL) {
+    OPENSSL_PUT_ERROR(RSA, ERR_R_PASSED_NULL_PARAMETER);
+    return NULL;
+  }
+
+  return rsa->meth;
+}
+
+void RSA_meth_free(RSA_METHOD *meth)
+{
+  if (meth != NULL) {
+    OPENSSL_free(meth);
+  }
+}
+
+int RSA_meth_set_init(RSA_METHOD *meth, int (*init) (RSA *rsa)) {
+  if(meth == NULL) {
+    OPENSSL_PUT_ERROR(RSA, ERR_R_PASSED_NULL_PARAMETER);
+    return 0;
+  }
+
+  meth->init = init;
+  return 1;
+}
+
+int RSA_meth_set_finish(RSA_METHOD *meth, int (*finish) (RSA *rsa)) {
+  if(meth == NULL) {
+    OPENSSL_PUT_ERROR(RSA, ERR_R_PASSED_NULL_PARAMETER);
+    return 0;
+  }
+
+  meth->finish = finish;
+  return 1;
+}
+
+int RSA_meth_set_priv_dec(RSA_METHOD *meth,
+                          int (*priv_dec) (int max_out, const uint8_t *from,
+                                           uint8_t *to, RSA *rsa,
+                                           int padding)) {
+  if(meth == NULL) {
+    OPENSSL_PUT_ERROR(RSA, ERR_R_PASSED_NULL_PARAMETER);
+    return 0;
+  }
+
+  meth->decrypt = priv_dec;
+  return 1;
+}
+
+int RSA_meth_set_priv_enc(RSA_METHOD *meth,
+                          int (*priv_enc) (int max_out, const uint8_t *from,
+                                           uint8_t *to, RSA *rsa,
+                                           int padding)) {
+  if(meth == NULL) {
+    OPENSSL_PUT_ERROR(RSA, ERR_R_PASSED_NULL_PARAMETER);
+    return 0;
+  }
+
+  meth->sign_raw = priv_enc;
+  return 1;
+}
+
+int RSA_meth_set_pub_dec(RSA_METHOD *meth,
+                         int (*pub_dec) (int max_out, const uint8_t *from,
+                                         uint8_t *to, RSA *rsa,
+                                         int padding)) {
+  if(meth == NULL) {
+    OPENSSL_PUT_ERROR(RSA, ERR_R_PASSED_NULL_PARAMETER);
+    return 0;
+  }
+
+  meth->verify_raw = pub_dec;
+  return 1;
+}
+
+int RSA_meth_set_pub_enc(RSA_METHOD *meth,
+                         int (*pub_enc) (int max_out, const uint8_t *from,
+                                         uint8_t *to, RSA *rsa,
+                                         int padding)) {
+  if(meth == NULL) {
+    OPENSSL_PUT_ERROR(RSA, ERR_R_PASSED_NULL_PARAMETER);
+    return 0;
+  }
+
+  meth->encrypt = pub_enc;
+  return 1;
+}
+
+int RSA_meth_set0_app_data(RSA_METHOD *meth, void *app_data) {
+  if(meth == NULL) {
+    OPENSSL_PUT_ERROR(RSA, ERR_R_PASSED_NULL_PARAMETER);
+    return 0;
+  }
+
+  meth->app_data = app_data;
+  return 1;
+}
+
 static int rsa_sign_raw_no_self_test(RSA *rsa, size_t *out_len, uint8_t *out,
                                      size_t max_out, const uint8_t *in,
                                      size_t in_len, int padding) {
   SET_DIT_AUTO_DISABLE;
   if (rsa->meth->sign_raw) {
```

Comment:
rsa->meth will never be NULL. It will either point to our default zero-initialized struct or to a custom one provided by the consumer. RSA_set_method does not allow for a NULL method to bet set and RSA_new always initializes new keys with a reference to the default struct. I'll add the check anyway just so we're more protected from future changes :)

Reply:
We have to include these checks in the internal functions. For example, rsa_sign_raw_no_self_test (our internal function) is called by the public RSA_sign_raw function. But that internal function is also called in the RSA_sign workflow (since sign wraps the lower level sign_raw similar to how in EC_KEY sign wrapped sign_sig).  If custom sign functionality is provided, RSA_sign will directly call that custom functionality. But in the case that custom sign functionality is not provided but custom sign_raw is, the RSA_sign workflow should call the lower level custom sign_raw which is only possible when this check is in the internal function.